fix: DIA-1907: [FE] TypeError: Failed to fetch

nick-skriabin opened the pull request:
Now handling the Failed to fetch error.

According to MDN it can happen due to network conditions or CORS error. In both cases we don't want to bother user with it, so we fail silently.

---

This pull request includes several important changes to the `useAbortController` hook and its usage across the codebase, as well as improvements to error handling in the `APIProxy` class.

### Changes to `useAbortController` Hook:

* **Refactored `useAbortController` Hook**: The `useAbortController` hook was moved from `web/apps/labelstudio/src/hooks/useAbortController.ts` to `web/libs/core/src/hooks/useAbortController.ts` and refactored to include an `abort` method for better error handling. [[1]](diffhunk://#diff-98547faed7cb056fc233b69511940666472d6f81da5267d02f675c3e89fcc3ecL1-L22) [[2]](diffhunk://#diff-86c7acfe099351a04a8b880df330b20f54baa287702f8b28e6b4a33b543779bbR1-R54)
* **Updated Imports**: Updated the import path for `useAbortController` in `Projects.jsx` to reflect the new location in `@humansignal/core`.
* **Export New Hook**: Added export for the refactored `useAbortController` in `web/libs/core/src/index.ts`.

### Improvements to Error Handling in `APIProxy`:

* **Ignored Errors**: Introduced a new regular expression `IGNORED_ERRORS` to catch and silently handle specific types of errors such as abort errors and network issues.
* **Enhanced Error Handling**: Modified the `APIProxy` class to use the `IGNORED_ERRORS` regex for improved error handling, ensuring certain errors do not reach the user.
<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

nick-skriabin committed:
Handle the Failed to fetch error

nick-skriabin committed:
Merge branch 'develop' into fb-dia-1907/failed-to-fetch

MihajloHoma commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14102052640)

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-dia-1907/failed-to-fetch'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14102052640

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14102052640

nick-skriabin committed:
Use shared `useAbortController`

MihajloHoma commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14130667682)

robot-ci-heartex committed:
Merge branch 'develop' into 'fb-dia-1907/failed-to-fetch'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14130667682

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14130667682

nick-skriabin committed:
Introducing `safeFetch`

Safe fetch will silence unwanted errors like AbortError or Failed to
fetch that are not supposed to be user facing. Instead, it will silently
report them as warning logs.

nick-skriabin committed:
Update APIProxy usage

nick-skriabin committed:
Fix DM not loading

hlomzik commented on web/libs/core/src/lib/utils/safe-fetch.ts:
```diff
@@ -0,0 +1,56 @@
+  if (error.message.match(IGNORED_ERRORS) !== null) {
+    return true;
+  }
+}
```
for consistent results
```suggestion
  return false;
}
```

nick-skriabin committed:
Revert "Fix DM not loading"

This reverts commit 0a58b7e62cbcb940751ef4fdf86832e554d5c14a.

nick-skriabin committed:
Revert "Update APIProxy usage"

This reverts commit 5eabcba599e7fda9e29be9b0d005124bba8604fb.

nick-skriabin committed:
Revert "Introducing `safeFetch`"

This reverts commit 60e3a1d62c9a466a058855805cd22e023cb95787.

bmartel commented on web/libs/core/src/hooks/useAbortController.ts:
```diff
@@ -0,0 +1,54 @@
+
+    if (ctrl.signal.aborted) return;
+
+    try {
+      ctrl.abort();
+    } catch (err) {
+      // handling errors during abort (especially "signal is aborted without a reason")
+      if (err instanceof Error && err.name.match(/abort/gi)) {
+        console.info(`AbortError: ${controller.current.signal.reason ?? "Operation aborted"}`);
+      } else {
+        console.error("Unexpected error during abort:", err);
+      }
+    }
```
This isn't needed as it won't ever throw here, the exception is raised in the original fetch call:

> When abort() is called, the fetch() promise rejects with an Error of type DOMException, with name AbortError

https://developer.mozilla.org/en-US/docs/Web/API/AbortController/abort#examples

nick-skriabin committed:
Remove unnecessary try-catch
